xbmc/guilib/GUIVisualisationControl.cpp

```diff
@@ -98,35 +151,324 @@ void CGUIVisualisationControl::Process(unsigned int currentTime, CDirtyRegionLis
+{
+  if (m_instance && m_alreadyStarted)
+    return m_instance->GetPresetList(vecpresets);
+
+  return false;
+}
+
+bool CGUIVisualisationControl::InitVisualization()
+{
+  ADDON::AddonPtr addonInfo;
+  ADDON::CAddonMgr::GetInstance().GetAddon(CServiceBroker::GetSettings().GetString(CSettings::SETTING_MUSICPLAYER_VISUALISATION), addonInfo, ADDON::ADDON_VIZ, true);
+  if (!addonInfo)
+    return false;
+
+  CServiceBroker::GetActiveAE().RegisterAudioCallback(this);
+
+  g_graphicsContext.CaptureStateBlock();
+
+  float x = g_graphicsContext.ScaleFinalXCoord(GetXPosition(), GetYPosition());
+  float y = g_graphicsContext.ScaleFinalYCoord(GetXPosition(), GetYPosition());
+  float w = g_graphicsContext.ScaleFinalXCoord(GetXPosition() + GetWidth(), GetYPosition() + GetHeight()) - x;
+  float h = g_graphicsContext.ScaleFinalYCoord(GetXPosition() + GetWidth(), GetYPosition() + GetHeight()) - y;
+  if (x < 0)
+    x = 0;
+  if (y < 0)
+    y = 0;
+  if (x + w > g_graphicsContext.GetWidth())
+    w = g_graphicsContext.GetWidth() - x;
+  if (y + h > g_graphicsContext.GetHeight())
+    h = g_graphicsContext.GetHeight() - y;
+
+  m_instance = new ADDON::CVisualization(std::dynamic_pointer_cast<ADDON::CAddonDll>(addonInfo), x, y, w, h);
+  CreateBuffers();
+
+  m_alreadyStarted = false;
+  g_graphicsContext.ApplyStateBlock();
+  return true;
+}
+
+void CGUIVisualisationControl::DeInitVisualization()
+{
+  if (!m_attemptedLoad)
+    return;
+
+  /*
+   * Prevent "UnregisterAudioCallback" in case Kodi is stopped. There becomes
+   * audio engine stopped before gui!
+   */
+  if (!g_application.m_bStop)
+    CServiceBroker::GetActiveAE().UnregisterAudioCallback(this);
+  
+  m_attemptedLoad = false;
+
   CGUIMessage msg(GUI_MSG_VISUALISATION_UNLOADING, m_controlID, 0);
   g_windowManager.SendMessage(msg);
+
   CLog::Log(LOGDEBUG, "FreeVisualisation() started");
-  CGUIRenderingControl::FreeResources(immediately);
-  m_addon.reset();
-  CLog::Log(LOGDEBUG, "FreeVisualisation() done");
+
+  if (m_instance)
+  {
+    if (m_alreadyStarted)
+    {
+      g_graphicsContext.CaptureStateBlock();
+      m_instance->Stop();
+      g_graphicsContext.ApplyStateBlock();
+      m_alreadyStarted = false;
+    }
+
+    delete m_instance;
+    m_instance = nullptr;
+  }
+
+  ClearBuffers();
+}
+
+void CGUIVisualisationControl::CreateBuffers()
+{
+  ClearBuffers();
+
+  // Get the number of buffers from the current vis
+  VIS_INFO info;
+
+  if (m_instance && m_alreadyStarted)
+    m_instance->GetInfo(&info);
+
+  m_numBuffers = info.iSyncDelay + 1;
+  m_wantsFreq = info.bWantsFreq;
+  if (m_numBuffers > MAX_AUDIO_BUFFERS)
+    m_numBuffers = MAX_AUDIO_BUFFERS;
+  if (m_numBuffers < 1)
+    m_numBuffers = 1;
 }
 
+void CGUIVisualisationControl::ClearBuffers()
+{
+  m_wantsFreq = false;
+  m_numBuffers = 0;
+
+  while (!m_vecBuffers.empty())
+  {
+    CAudioBuffer* pAudioBuffer = m_vecBuffers.front();
+    delete pAudioBuffer;
+    m_vecBuffers.pop_front();
+  }
```

Comment:
Once switch to a vector of unique_ptr, this whole block can just become `m_vecBuffers.clear()`